[DOCS] Adding create index snapshot API docs (#58519)
elastic/elasticsearch: docs/reference/snapshot-restore/apis/create-snapshot-api.asciidoc

> [[create-snapshot-api]]
> === Create index snapshot API
> ++++
> <titleabbrev>Create index snapshot</titleabbrev>
> ++++

Reviewer: By default, snapshots include an entire cluster, including the cluster state.
With #57675, snapshots now support data streams in `master` and `7.x`.

I'd avoid associating snapshots too closely with indices.

```suggestion
[[create-snapshot-api]]
=== Create snapshot API
++++
<titleabbrev>Create snapshot</titleabbrev>
++++
```

Takes a snapshot of one or more indices. A repository can contain multiple snapshots of the same cluster. Snapshots are identified by unique names in the cluster, such as `snapshot_1`.

////
[source,console]
-----------------------------------
PUT /_snapshot/my_backup
{
  "type": "fs",
  "settings": {
    "location": "my_backup_location"
  }
}
-----------------------------------
// TESTSETUP
////

[source,console]
-----------------------------------

> PUT /_snapshot/my_backup/snapshot_1?wait_for_completion=true

Author: I was curious about whether to include the parameter, so thank you for the guidance.

-----------------------------------

[[create-snapshot-api-request]]
==== {api-request-title}

`POST /_snapshot/<repository>/<snapshot-name>`

[[create-snapshot-api-desc]]
==== {api-description-title}

> By default, a snapshot of all open and started indices in the cluster is created. You can change this behavior by specifying the list of indices in the body of the snapshot request.

Reviewer: I would add a higher-level intro sentence. We also need to mention data streams.
```suggestion
You can use the create snapshot API to create a <<snapshot-restore,snapshot>>, a
backup taken from a running {es} cluster.

By default, a snapshot includes all data streams and open indices in the
cluster, as well as the cluster state.  You can change this behavior by
specifying a list of data streams and indices to back up in the body of the
snapshot request.
```

NOTE: You must register a snapshot before performing snapshot and restore operations. Use the <<put-snapshot-repo-api,put snapshot repository API>> to register new repositories and update existing ones.

> The index snapshot process is incremental. When creating the index snapshot, {es} analyzes the list of index files that are already stored in the repository and copies only files that were created or changed since the last snapshot. This process allows multiple snapshots to be preserved in the repository in a compact form.

Reviewer: I made some suggestions to remove index references. These also apply to data streams now.
```suggestion
The snapshot process is incremental. When creating a  snapshot, {es} analyzes the list of files that are already stored in the repository and copies only files that were created or changed since the last snapshot. This process allows multiple snapshots to be preserved in the repository in a compact form.
```

> The snapshot process is executed in non-blocking fashion, so all indexing and searching operations can run concurrently against the index that {es} is snapshotting. Only one snapshot process can run in the cluster at any time.

Reviewer: ```suggestion
The snapshot process is executed in non-blocking fashion, so all indexing and searching operations can run concurrently against the data stream or index that {es} is snapshotting. Only one snapshot process can run in the cluster at any time.
```

> Because a snapshot represents the index at the moment when the snapshot was created, no records that were added to the index after the snapshot process started will be present in the snapshot.

Reviewer: ```suggestion
A snapshot represents a point-in-time view of the moment when the snapshot was created. No records that were added to a data stream or index after the snapshot process started will be present in the snapshot.
```

For primary shards that have not been started and are not currently relocating, the snapshot process starts immediately. If shards are in the process of starting or relocating, {es} waits for these processes to complete before taking a snapshot.

IMPORTANT: While a snapshot of a particular shard is being created, this shard cannot be moved to another node. Relocating a shard during the snapshot process can interfere with rebalancing and allocation filtering. {es} can move a shard to another node (according to the current allocation filtering settings and rebalancing algorithm) only after the snapshot process completes.

> Besides creating a copy of each index, the snapshot process can also store global cluster metadata, including persistent cluster settings and templates. The transient settings and registered snapshot repositories are not stored as part of the snapshot.

Reviewer: ```suggestion
Besides creating a copy of each data stream and index, the snapshot process can also store global cluster metadata, including persistent cluster settings and templates. The transient settings and registered snapshot repositories are not stored as part of the snapshot.
```

[[create-snapshot-api-path-params]]
==== {api-path-parms-title}

`<repository>`::
(Required, string)
Name of the repository to create a snapshot in.

`<snapshot-name>`::
(Required, string)
Unique name of the snapshot.

> `wait_for_completion`::

Author: I'll move `wait_for_completion` to the query params section.

(Optional, boolean)
Specifies whether the request returns immediately after snapshot initialization (default), or wait for snapshot completion. Defaults to `false`.
+
NOTE: During snapshot initialization, information about all
previous snapshots is loaded into memory. In large repositories, this load time can cause this command to take several seconds (or even minutes) to return, even if the `wait_for_completion` parameter is set to `false`.

[[create-snapshot-api-query-params]]

> ==== {api-query-parms-title}

Author: I'll make this change to request body params and also reorder them.

`indices`::
(Optional, string)
A list of indices to include in the snapshot. This parameter supports <<multi-index,multi index syntax>>, although the options that control the behavior of multi index syntax
must be supplied in the body of the request, rather than as request parameters.

include::{es-repo-dir}/rest-api/common-parms.asciidoc[tag=index-ignore-unavailable]

`include_global_state`::
(Optional, boolean)
If `false`, this parameter prevents the cluster global state from being stored as part of the snapshot. Defaults to `false`.
+
If `true`, the snapshot's persistent settings, index templates, ingest
pipelines, and {ilm-init} policies are restored into the current cluster. This
overwrites any existing cluster settings, templates, pipelines and {ilm-init}
policies whose names match those in the snapshot.
+

> IMPORTANT: By default, the entire snapshot will fail if one or more indices participating in the snapshot do not have all primary shards available. You can change this behavior by setting `partial` to `true`.

Author: I'll add `partial` to the list of query parameters.

`metadata`::
(Optional, string)

> Attaches arbitrary metadata to the snapshot, such as a record of who took the snapshot, why it was taken, or any other useful data.

Author: Adding to the `metadata` description:
>Metadata must be less than 1024 bytes.

> 

Author: I'll add a description of `settings` with a pointer to that information in the [Put snapshot repository API docs](https://www.elastic.co/guide/en/elasticsearch/reference/current/put-snapshot-repo-api.html).

[[create-snapshot-api-example]]
==== {api-examples-title}

In the following example, `index_1` and `index_2` are the indices to take snapshots of.

[source,console]
-----------------------------------
PUT /_snapshot/my_backup/snapshot_2?wait_for_completion=true
{
  "indices": "index_1,index_2",
  "ignore_unavailable": true,
  "include_global_state": false,
  "metadata": {
    "taken_by": "user123",
    "taken_because": "backup before upgrading"
  }
}
-----------------------------------

The API returns the following response:

> [source,console-result]
> ----
> {
>   "accepted" : true
> }
> ----
> // TEST[skip:cannot complete subsequent snapshot]

Reviewer: Snippet tests help ensure our docs are correct. Generally, we don't use `// TEST[skip:...` unless absolutely necessary.  In this case,  I think the skipped test is hiding an incorrect response.

I suggested a valid response below. Here's how you can get it yourself next time:

* Use a command like `./gradlew :docs:integTestRunner --tests "*create-snapshot-api*"` to test the snippets in the file.
* Use `// TESTRESPONSE[s/.../.../]` to selectively replace some properties in the response.

This is covered in more depth in the docs README: https://github.com/elastic/elasticsearch/tree/master/docs

```suggestion
[source,console-result]
----
{
  "snapshot": {
    "snapshot": "snapshot_2",
    "uuid": "vdRctLCxSketdKb54xw67g",
    "version_id": 8000099,
    "version": "8.0.0",
    "indices": [],
    "data_streams": [],
    "include_global_state": false,
    "metadata": {
      "taken_by": "user123",
      "taken_because": "backup before upgrading"
    },
    "state": "SUCCESS",
    "start_time": "2020-06-25T14:00:28.850Z",
    "start_time_in_millis": 1593093628850,
    "end_time": "2020-06-25T14:00:28.850Z",
    "end_time_in_millis": 1593094752018,
    "duration_in_millis": 0,
    "failures": [],
    "shards": {
      "total": 0,
      "failed": 0,
      "successful": 0
    }
  }
}
----
// TESTRESPONSE[s/"uuid": "vdRctLCxSketdKb54xw67g"/"uuid": $body.snapshot.uuid/]
// TESTRESPONSE[s/"start_time": "2020-06-25T14:00:28.850Z"/"start_time": $body.snapshot.start_time/]
// TESTRESPONSE[s/"start_time_in_millis": 1593093628850/"start_time_in_millis": $body.snapshot.start_time_in_millis/]
// TESTRESPONSE[s/"end_time": "2020-06-25T14:00:28.850Z"/"end_time": $body.snapshot.end_time/]
// TESTRESPONSE[s/"end_time_in_millis": 1593094752018/"end_time_in_millis": $body.snapshot.end_time_in_millis/]
// TESTRESPONSE[s/"duration_in_millis": 0/"duration_in_millis": $body.snapshot.duration_in_millis/]
```

[[create-snapshot-api]]
=== Create snapshot API
++++
<titleabbrev>Create snapshot</titleabbrev>
++++

Takes a <<snapshot-restore,snapshot>> of a cluster or specified data streams and
indices.

////
[source,console]
-----------------------------------
PUT /_snapshot/my_repository
{
  "type": "fs",
  "settings": {
    "location": "my_backup_location"
  }
}
-----------------------------------
// TESTSETUP
////

> [source,console]
> -----------------------------------
> PUT /_snapshot/my_repository/my_snapshot
> -----------------------------------

Reviewer: It looks like the docs CI is failing because this snapshot doesn't finish before the example at the bottom. You can spoof the `wait_for_completion` query param using a `// TEST` comment.
```suggestion
[source,console]
-----------------------------------
PUT /_snapshot/my_repository/my_snapshot
-----------------------------------
// TEST[s/my_snapshot/my_snapshot?wait_for_completion=true/]
```

Sets snapshot settings for the repository in JSON or YAML format. See <<put-snapshot-repo-api-settings-param,`settings`>> in the Put snapshot repository API documentation.

`wait_for_completion`::
(Optional, boolean)
If `true`, the request returns a response when the snapshot is complete.
If `false`, the request returns a response when the snapshot initializes.
Defaults to `false`.
+
NOTE: During snapshot initialization, information about all
previous snapshots is loaded into memory. In large repositories, this load time can cause requests to take several seconds (or even minutes) to return a response, even if the `wait_for_completion` parameter is `false`.

[[create-snapshot-api-example]]
==== {api-examples-title}

The following request takes a snapshot of `index_1` and `index_2`.

[source,console]
-----------------------------------
PUT /_snapshot/my_repository/snapshot_2?wait_for_completion=true
{
  "indices": "index_1,index_2",
  "ignore_unavailable": true,
  "include_global_state": false,
  "metadata": {
    "taken_by": "user123",
    "taken_because": "backup before upgrading"
  }
}
-----------------------------------

The API returns the following response:

> [source,console-result]
> ----
> {
>   "snapshot": {
>     "snapshot": "snapshot_2",
>     "uuid": "vdRctLCxSketdKb54xw67g",
>     "version_id": 8000099,
>     "version": "8.0.0",
>     "indices": [],
>     "data_streams": [],
>     "include_global_state": false,
>     "metadata": {
>       "taken_by": "user123",
>       "taken_because": "backup before upgrading"
>     },
>     "state": "SUCCESS",
>     "start_time": "2020-06-25T14:00:28.850Z",
>     "start_time_in_millis": 1593093628850,
>     "end_time": "2020-06-25T14:00:28.850Z",
>     "end_time_in_millis": 1593094752018,
>     "duration_in_millis": 0,
>     "failures": [],
>     "shards": {
>       "total": 0,
>       "failed": 0,
>       "successful": 0
>     }
>   }
> }
> ----
> // TESTRESPONSE[s/"uuid": "vdRctLCxSketdKb54xw67g"/"uuid": $body.snapshot.uuid/]

Reviewer: I didn't notice this earlier, but it looks like this response includes a `version_id` and `version`. This won't cause problems on the `master` branch, but it'll fail as soon as it's ported to any other branches.

The suggestion below replaces the version properties with variables and fixes them with `// TESTRESPONSE` comments.
```suggestion
[source,console-result]
----
{
  "snapshot": {
    "snapshot": "snapshot_2",
    "uuid": "vdRctLCxSketdKb54xw67g",
    "version_id": <version_id>,
    "version": <version>,
    "indices": [],
    "data_streams": [],
    "include_global_state": false,
    "metadata": {
      "taken_by": "user123",
      "taken_because": "backup before upgrading"
    },
    "state": "SUCCESS",
    "start_time": "2020-06-25T14:00:28.850Z",
    "start_time_in_millis": 1593093628850,
    "end_time": "2020-06-25T14:00:28.850Z",
    "end_time_in_millis": 1593094752018,
    "duration_in_millis": 0,
    "failures": [],
    "shards": {
      "total": 0,
      "failed": 0,
      "successful": 0
    }
  }
}
----
// TESTRESPONSE[s/"uuid": "vdRctLCxSketdKb54xw67g"/"uuid": $body.snapshot.uuid/]
// TESTRESPONSE[s/"version_id": <version_id>/"version_id": $body.snapshot.version_id/]
// TESTRESPONSE[s/"version": <version>/"version": $body.snapshot.version/]
```

indices.

////
[source,console]
-----------------------------------
PUT /_snapshot/my_repository
{
  "type": "fs",
  "settings": {
    "location": "my_backup_location"
  }
}
-----------------------------------
// TESTSETUP
////

[source,console]
-----------------------------------
PUT /_snapshot/my_repository/my_snapshot
-----------------------------------
// TEST[s/my_snapshot/my_snapshot?wait_for_completion=true/]

[[create-snapshot-api-request]]
==== {api-request-title}

`PUT /_snapshot/<repository>/<snapshot>`

`POST /_snapshot/<repository>/<snapshot>`

[[create-snapshot-api-desc]]
==== {api-description-title}

You can use the create snapshot API to create a <<snapshot-restore,snapshot>>, which is a
backup taken from a running {es} cluster.

By default, a snapshot includes all data streams and open indices in the
cluster, as well as the cluster state.  You can change this behavior by
specifying a list of data streams and indices to back up in the body of the
snapshot request.

NOTE: You must register a snapshot before performing snapshot and restore operations. Use the <<put-snapshot-repo-api,put snapshot repository API>> to register new repositories and update existing ones.

> The snapshot process is incremental. When creating a  snapshot, {es} analyzes the list of files that are already stored in the repository and copies only files that were created or changed since the last snapshot. This process allows multiple snapshots to be preserved in the repository in a compact form.

Reviewer: Extra space. Line wrapping is also too long.
```suggestion
The snapshot process is incremental. When creating a snapshot, {es} analyzes the list of files that are already stored in the repository and copies only files that were created or changed since the last snapshot. This process allows multiple snapshots to be preserved in the repository in a compact form.
```

`include_global_state`::
+
--
(Optional, boolean)
If `true`, the current cluster state is included in the snapshot.
Defaults to `true`.

The cluster state includes:

* Persistent cluster settings
* Index templates
* Legacy index templates
* Ingest pipelines
* {ilm-init} lifecycle policies
--
+
IMPORTANT: By default, the entire snapshot will fail if one or more indices included in the snapshot do not have all primary shards available. You can change this behavior by setting <<create-snapshot-api-partial,`partial`>> to `true`.

include::{es-repo-dir}/rest-api/common-parms.asciidoc[tag=master-timeout]

`metadata`::
(Optional, string)
Attaches arbitrary metadata to the snapshot, such as a record of who took the snapshot, why it was taken, or any other useful data. Metadata must be less than 1024 bytes.

[[create-snapshot-api-partial]]
`partial`::
(Optional, boolean)
If `false`, the entire snapshot will fail if one or more indices included in the snapshot do not have all primary shards available. Defaults to `false`.
+
If `true`, allows taking a partial snapshot of indices with unavailable shards.

`settings`::
(Optional, object)

> Sets snapshot settings for the repository in JSON or YAML format. See <<put-snapshot-repo-api-settings-param,`settings`>> in the Put snapshot repository API documentation.

Reviewer: I tested this parameter, and it seems weird... you can throw any arbitrary object in there and it accepts it.

@original-brownbear Any thoughts? I wonder if this is just a param that never got used. If so, we should maybe look at removing it.

In any case, I don't think this definition is accurate or makes sense. You shouldn't be able to overwrite snapshot repo settings from a snapshot.

If this is just an unused param, I wonder if we should A) not document it or B) document it with a warning that it doesn't do anything. I'd hate to see a user throw values in here expecting something to happen if it does nothing.

Reviewer 2: > If this is just an unused param, I wonder if we should A) not document it or B) document it with a warning that it doesn't do anything. I'd hate to see a user throw values in here expecting something to happen if it does nothing.

Sorry for the delay here. You're right ... this parameter doesn't do anything at the moment (I'll look into removing it). IMO, we probably should just not document it. Was it ever documented in the first place even?

Reviewer 2: https://github.com/elastic/elasticsearch/pull/58661

Author: Thanks @jrodewig for testing this parameter, and @original-brownbear for confirming that we can remove it. With that parameter removed, I think that this PR is ready to merge.

Reviewer: Agree. Thanks for taking this one to the finish line @lockewritesdocs.
